File: backend/src/models/Leave.js
import mongoose from "mongoose";
import { LEAVE_TYPES, LEAVE_REQUEST_STATUS } from "../config/constants.js";

const leaveSchema = new mongoose.Schema(
  {
    employeeId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Employee",
      required: [true, "Employee is required"],
    },
    leaveType: {
      type: String,
      required: [true, "Leave type is required"],
      enum: {
        values: LEAVE_TYPES,
        message:
          "Leave type must be one of: SICK, CASUAL, ANNUAL, UNPAID, MATERNITY, PATERNITY",
      },
    },
    startDate: {
      type: Date,
      required: [true, "Start date is required"],
    },
    endDate: {
      type: Date,
      required: [true, "End date is required"],
    },
    days: {
      type: Number,
      required: [true, "Number of days is required"],
      min: [1, "Leave must be at least 1 day"],
    },
    reason: {
      type: String,
      trim: true,
    },
    status: {
      type: String,
      enum: {
        values: LEAVE_REQUEST_STATUS,
        message: "Status must be PENDING, APPROVED, REJECTED, or CANCELLED",
      },
      default: "PENDING",
    },
    approvedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Employee",
    },
    approvedAt: {
      type: Date,
    },
  },
  {
    timestamps: true,
  },
);

export default mongoose.model("Leave", leaveSchema);

File: frontend/lib/models/leave.dart
class Leave {
  final String id;
  final String employeeId;
  final String leaveType;
  final DateTime startDate;
  final DateTime endDate;
  final int days;
  final String? reason;
  final String status;

  Leave({
    required this.id,
    required this.employeeId,
    required this.leaveType,
    required this.startDate,
    required this.endDate,
    required this.days,
    this.reason,
    required this.status,
  });

  factory Leave.fromJson(Map<String, dynamic> json) {
    return Leave(
      id: json['_id'] ?? json['id'] ?? '',
      employeeId: json['employeeId'] ?? '',
      leaveType: json['leaveType'] ?? 'CASUAL',
      startDate: DateTime.parse(
        json['startDate'] ?? DateTime.now().toIso8601String(),
      ),
      endDate: DateTime.parse(
        json['endDate'] ?? DateTime.now().toIso8601String(),
      ),
      days: json['days'] ?? 1,
      reason: json['reason'],
      status: json['status'] ?? 'PENDING',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'employeeId': employeeId,
      'leaveType': leaveType,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'days': days,
      'reason': reason,
    };
  }
}
